File: BackEnd/src/data/IMappers/IReminderMapper.ts
/**
 * @file IReminderMapper.ts
 * @description Interfaz del mapper de recordatorios de riego.
 * @module Reminders
 * @layer Data
 */

import type { Reminder } from '../../domain/entities/Reminder.js';
import type { ReminderResponseDTO } from '../../domain/dtos/reminders/reminder-response.dto.js';
import type { ReminderDocument } from '../datasources/mongodb/models/ReminderModel.js';

export interface IReminderMapper {
  /**
   * Convierte un documento MongoDB en entidad de dominio.
   *
   * @param doc — Documento de la colección reminders.
   * @returns Entidad Reminder.
   */
  toEntity(doc: ReminderDocument): Reminder;

  /**
   * Convierte una entidad en documento listo para MongoDB.
   *
   * @param entity — Entidad Reminder.
   * @returns ReminderDocument.
   */
  toDocument(entity: Reminder): ReminderDocument;

  /**
   * Convierte una entidad en DTO de respuesta HTTP.
   *
   * @param entity — Entidad Reminder.
   * @returns ReminderResponseDTO.
   */
  toResponseDTO(entity: Reminder): ReminderResponseDTO;
}

File: FrontEnd/lib/data/i_mappers/i_reminder_mapper.dart
/// @file i_reminder_mapper.dart
/// @description Interfaz del mapper de recordatorios. Contrato ReminderModel ↔ Reminder.
/// @module Reminders
/// @layer Data
library;

import '../../domain/entities/reminder.dart';
import '../models/reminder_model.dart';

/// Contrato de conversión entre el modelo de serialización y la entidad de dominio.
///
/// [injectable] registrar en container.dart como singleton.
abstract interface class IReminderMapper {
  /// Convierte un modelo de serialización a entidad de dominio.
  Reminder toEntity(ReminderModel model);

  /// Convierte una entidad de dominio a modelo de serialización.
  ReminderModel toModel(Reminder entity);
}
